ldap/servers/slapd/csngen.c

```diff
@@ -313,79 +292,60 @@ csngen_adjust_time(CSNGen *gen, const CSN *csn)
-        time_t new_time = 0;
-
-        if (remote_seqnum > gen->state.seq_num) {
-            if (remote_seqnum < CSN_MAX_SEQNUM) {
-                gen->state.seq_num = remote_seqnum + 1;
-            } else {
-                remote_time++;
-            }
-        }
-
-        remote_offset = remote_time - cur_time;
-        if (remote_offset > gen->state.remote_offset) {
-            if (ignore_time_skew || (remote_offset <= CSN_MAX_TIME_ADJUST)) {
-                gen->state.remote_offset = remote_offset;
-            } else /* remote_offset > CSN_MAX_TIME_ADJUST */
-            {
-                slapi_log_err(SLAPI_LOG_ERR, "csngen_adjust_time",
-                              "Adjustment limit exceeded; value - %ld, limit - %ld\n",
-                              remote_offset, (long)CSN_MAX_TIME_ADJUST);
-                slapi_rwlock_unlock(gen->lock);
-                csngen_dump_state(gen, SLAPI_LOG_DEBUG);
-                return CSN_LIMIT_EXCEEDED;
-            }
-        } else if (remote_offset > 0) { /* still need to account for this */
-            gen->state.local_offset += remote_offset;
+    remote_offset = remote_time - CSN_CALC_TSTAMP(gen);
+    if (remote_offset > 0) {
+        if (!ignore_time_skew && (gen->state.remote_offset + remote_offset > CSN_MAX_TIME_ADJUST)) {
+            slapi_log_err(SLAPI_LOG_ERR, "csngen_adjust_time",
+                          "Adjustment limit exceeded; value - %ld, limit - %ld\n",
+                          remote_offset, (long)CSN_MAX_TIME_ADJUST);
+            slapi_rwlock_unlock(gen->lock);
+            csngen_dump_state(gen, SLAPI_LOG_DEBUG);
+            return CSN_LIMIT_EXCEEDED;
         }
-
-        new_time = CSN_CALC_TSTAMP(gen);
-        /* let's revisit the seq num - if the new time is > the old
-           tiem, we should reset the seq number to remote + 1 if
-           this won't cause a wrap around */
-        if (new_time >= cur_time) {
-            /* just set seq_num regardless of whether the current one
-               is < or > than the remote one - the goal of this function
-               is to make sure we generate CSNs > the remote CSN - if
-               we have increased the time, we can decrease the seqnum
-               and still guarantee that any new CSNs generated will be
-               > any current CSNs we have generated */
-            if (remote_seqnum < gen->state.seq_num) {
-                gen->state.seq_num ++;
-            } else {
-                gen->state.seq_num = remote_seqnum + 1;
-            }
+        gen->state.remote_offset += remote_offset;
+        /* To avoid beat phenomena between suppliers let put 1 second in local_offset
+         * it will be eaten at next clock tick rather than increasing remote offset
+         * If we do not do that we will have a time skew drift of 1 second per 2 seconds
+         * if suppliers are desynchronized by 0.5 second 
+         */
+        if (gen->state.local_offset == 0) {
```

Comment:
It increase local_offset by 1 and decrease remote_offset by 1 only if local_offset == 0
Why not systematically ?